### PR TITLE
[UI-22] Build components for common charts

### PR DESCRIPTION
**Pull request by maciek15**
#### Task

- **Issue or task referred:** UI-22
- **Feature or Bugfix:** Feature

I have prepared components for common charts (Bar, Line and Pie) and Legend for charts. Additionally those components are available with filters.

#### Checklist

- [x] Required unit tests prepared
- [x] Documentation prepared
- [x] External licenses validated (ideally MIT)

### CODE REVIEW

File: packages/charts/README.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+className     | string          | n/a                   | Additional class name passed to wrapper.
+data          | ChartData       | `[{ dataItems: [] }]` | Line chart data.
+timeSeries    | boolean         | `false`               | Idicates if x axis should be displayed as date 
+type          | `line` or `bar` | `line`                | Type of chart (line or bar).
+xAxisTitle    | string          | n/a                   | X axis title.
+yAxisTitle    | string          | n/a                   | Y axis title.
+zoomable      | boolean         | `false`               | Indicates if chart is zoomable, can be applied to chart.
+
+
+### ChartWithFilter
```
Comment (rangoo94) [inline packages/charts/README.md]: Isn't there HOC instead?

File: packages/charts/README.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+- `prop-types: ^15.6.1`
+- `react: ^16.2.0`
+- `react-dom: ^16.2.0`
+
+These packages are required by `@talixo/charts`, but you have to install them manually,
+to avoid having different versions of these in your application.
+
+## Supported props
+
+### Chart
```
Comment (rangoo94) [inline packages/charts/README.md]: You haven't described `Highlight`

File: packages/charts/src/Chart.js [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+Chart.propTypes = propTypes
+
+Chart.defaultProps = defaultProps
+
+Chart.displayName = 'Chart'
+
+export default Chart
+export const ChartWithFilters = WithFilters(Chart)
```
Comment (rangoo94) [inline packages/charts/src/Chart.js]: I think that you don't have to export that

File: packages/charts/src/Highlight.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
+import React from 'react'
+import { ScaleUtils, AbstractSeries } from 'react-vis'
+
+/**
+ * UI component that represents hightlighted drawing area above a Chart component
```
Comment (rangoo94) [inline packages/charts/src/Highlight.js]: Typo `highlighted`

File: packages/charts/src/Highlight.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+      left: xScale.invert(drawArea.left)
+    }
+
+    if (onBrushEnd) {
+      onBrushEnd(domainArea)
+    }
+  }
+
+  onParentMouseMove (e) {
+    const {marginLeft, onBrush} = this.props
```
Comment (rangoo94) [inline packages/charts/src/Highlight.js]: Once it is:
```js
const drawArea = { top: 0, right: 0, bottom: 0, left: 0 }
const { marginLeft, innerHeight, onBrushStart } = this.props
const { onBrushEnd } = this.props
const { drawArea } = this.state
```
and once:
```js
const {marginLeft, onBrush} = this.props
const {drawing} = this.state
this.setState({drawArea: newDrawArea})
```
could you keep consistent style at least across file? :)

File: packages/charts/src/Highlight.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+      this.setState({drawArea: newDrawArea})
+
+      if (onBrush) {
+        onBrush(e)
+      }
+    }
+  }
+
+  render () {
+    const {
```
Comment (rangoo94) [inline packages/charts/src/Highlight.js]: Are you sure that you want to keep one per line? ;) It's making `render` function much bigger without reason

File: packages/charts/src/Highlight.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+          height={bottom}
+        />
+      </g>
+    )
+  }
+}
+
+Highlight.displayName = 'HighlightOverlay'
+
+Highlight.defaultProps = {
```
Comment (rangoo94) [inline packages/charts/src/Highlight.js]: You have passed `defaultProps`, but what about `propTypes`?

File: packages/charts/src/Legend.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    color: PropTypes.string,
+
+    /** Indicates if series is disabled. */
+    disabled: PropTypes.bool,
+
+    /** Data series label. */
+    label: PropTypes.string,
+
+    /** Data series value. */
+    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
```
Comment (rangoo94) [inline packages/charts/src/Legend.js]: Again, once it is:
```js
[PropTypes.string, PropTypes.number]
```
and once:
```js
[ direction ]
```

We have to force consistent style in ESLint.

File: packages/charts/styles/config.sass [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+      fill: $color!important
+
+@mixin graph-colors ($baseColor)
+  @for $i from 0 to 5
+    $multiplier: $i * 10%
+    @include graph-color($i, darken($baseColor, $multiplier))
+
+  @for $i from 0 to 5
+    $multiplier:  $i * 10%
+    @include graph-color($i + 5, lighten($baseColor, $multiplier))
```
Comment (rangoo94) [inline packages/charts/styles/config.sass]: Right now base color is used for two groups, and additionally, it's not smooth change (shouldn't it be firstly `lighten` and later `darken`?). Example: right now, for color #999:

```css
.color--0 { background-color: #999999; }
.color--1 { background-color: gray; }
.color--2 { background-color: #666666; }
.color--3 { background-color: #4d4d4d; }
.color--4 { background-color: #333333; }
.color--5 { background-color: #999999; }
.color--6 { background-color: #b3b3b3; }
.color--7 { background-color: #cccccc; }
.color--8 { background-color: #e6e6e6; }
.color--9 { background-color: white; }
```
So, as you can see, up to `.color--4` it's getting darker, when immediately in `.color--5` is much lighter (same as first element), and later is going lighter and lighter. Maybe even `10%` is too much, when `color` is going from `#999` to `#fff`

File: packages/charts/styles/geometry.sass [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  right: 0
+  box-sizing: border-box
+  padding: $legend-padding
+
+  &--horizontal
+    max-width: 100%
+    flex-direction: row
+    overflow-x: auto
+
+    #{prefix(legend-item)}
```
Comment (rangoo94) [inline packages/charts/styles/geometry.sass]: You have forgotten quotes ;)

File: packages/charts/styles/geometry.sass [hunk 11]
```diff
@@ -0,0 +1,73 @@
+@import './config'
+
+@include module('pie-chart')
```
Comment (rangoo94) [inline packages/charts/styles/geometry.sass]: Is this empty module needed?

File: packages/charts/styles/geometry.sass [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      white-space: nowrap
+
+      &__colorbox
+        display: block
+
+  &--vertical
+    flex-direction: column
+    overflow-y: auto
+
+    #{prefix(legend-item)}
```
Comment (rangoo94) [inline packages/charts/styles/geometry.sass]: Forgotten quotes

File: packages/charts/src/Highlight.js [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+ * UI component that represents highlighted drawing area above a Chart component
+ * This is a copy of solution provided by react-vis team
+ * https://github.com/uber/react-vis/blob/master/showcase/examples/zoomable-chart/highlight.js
+ *
+ * @property {object} [props]
+ * @property {string} [props.color]
+ * @property {nomber} [props.innerHeight]
+ * @property {nomber} [props.innerWidth]
+ * @property {nomber} [props.marginLeft]
+ * @property {nomber} [props.marginTop]
```
Comment (PiotrGrzechnik) [inline packages/charts/src/Highlight.js]: Typo mistake `number`

File: packages/charts/src/Highlight.js [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+ * @property {function} [props.onBrush]
+ * @property {function} [props.onBrushStart]
+ * @property {function} [props.onBrushEnd]
+ * @property {nomber} [props.opacity]
```
Comment (PiotrGrzechnik) [inline packages/charts/src/Highlight.js]: The same as above ;)

File: packages/charts/src/utils.js [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    .map(item => item.value)
+    .reduce((acc, curr) => acc + curr, 0)
+}
+
+/**
+ * Generates data for arc series
+ * @param {array} data
+ * @param {number} sum
+ * @param {object} config
+ * @returns {arrat}
```
Comment (PiotrGrzechnik) [inline packages/charts/src/utils.js]: `array` ;)

File: packages/charts/styles/theming.sass [hunk 16]
```diff
@@ -0,0 +1,23 @@
+@import './config'
+
+@include graph-colors($color-legend-base)
+
+@include module('legend')
+  background-color: $legend-background
+
+
```
Comment (PiotrGrzechnik) [inline packages/charts/styles/theming.sass]: empty line

File: packages/charts/styles/config.sass [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+@mixin graph-colors ($baseColor)
+  @for $i from 0 to 5
+    $multiplier: (5 - $i) * 7%
+    @include graph-color($i, lighten($baseColor, $multiplier))
+
+  @for $i from 0 to 5
+    $multiplier:  ($i + 1) * 7%
+    @include graph-color($i + 5, darken($baseColor, $multiplier))
+
+
```
Comment (PiotrGrzechnik) [inline packages/charts/styles/config.sass]: empty line

File: packages/charts/README.md (hunk 2 above)
Comment (maciek15) [reply packages/charts/README.md]: Hightlight is not exported

File: packages/charts/package.json [hunk 18]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "@talixo/charts",
+  "version": "1.0.0",
```
Comment (stuniel) [inline packages/charts/package.json]: We can change the version to `0.1.0`.

File: packages/charts/src/Chart.js [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+
+import { buildClassName } from '@talixo/shared'
+
+import Highligth from './Highlight'
+import { generateSeriesClassName } from './utils'
+
+const moduleName = 'chart'
+
+const Components = {
+  bar: VerticalBarSeries.displayName = 'BarChart' && VerticalBarSeries,
```
Comment (rangoo94) [inline packages/charts/src/Chart.js]: `VerticalBarSeries.displayName = 'BarChart'` should be in VerticalBarSeries, or different way

File: packages/charts/src/Chart.js [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+    PropTypes.shape({
+      /** Additional data series className. */
+      className: PropTypes.string,
+
+      /** Color of the data series. */
+      color: PropTypes.string,
+
+      /** Data information to be displayed in chart. */
+      dataItems: PropTypes.arrayOf(PropTypes.shape({
+
```
Comment (rangoo94) [inline packages/charts/src/Chart.js]: unnecessary empty line

File: packages/charts/src/Chart.js [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+  /** Y axis title. */
+  yAxisTitle: PropTypes.string,
+
+  /** Indicates if chart is zoomable, can be applied to chart. */
+  zoomable: PropTypes.bool
+}
+
+const defaultProps = {
+  data: [{ dataItems: [] }],
```
Comment (rangoo94) [inline packages/charts/src/Chart.js]: shouldn't be empty at all?

File: packages/charts/src/Chart.js [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+                  animate
+                  style={isLineChart ? {fill: 'none'} : {}}
+                  {...getSeriesProps(item, index)}
+                />
+                : null
+            })
+        }
+        {
+          zoomable && type === 'line' &&
+          <Highligth
```
Comment (rangoo94) [inline packages/charts/src/Chart.js]: `Highlight`

File: packages/charts/src/Legend.js [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+ * @param {string} [props.dataItems.color]
+ * @param {bool} [props.dataItems.disabled]
+ * @param {string} [props.dataItems.label]
+ * @param {string|number} [props.dataItems.value]
+ * @param {string} [props.direction]
+ * @param {function} [props.onClick]
+ * @returns {React.Element}
+ */
+function Legend (props) {
+  const {
```
Comment (rangoo94) [inline packages/charts/src/Legend.js]: Single line?

File: packages/charts/src/LegendItem.js [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+ * @param {string} [props.color]
+ * @param {bool} [props.disabled]
+ * @param {object} [props.id]
+ * @param {function} [props.onClick]
+ * @param {string} [props.label]
+ * @param {string|number} [props.value]
+ * @returns {ReactElement}
+ */
+function LegendItem (props) {
+  const {
```
Comment (rangoo94) [inline packages/charts/src/LegendItem.js]: Single line?

File: packages/charts/src/PieChart.js [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+const propTypes = {
+  /** Additional props that are passed to react-vis ArcSeries. */
+  arcProps: PropTypes.object,
+
+  /** Additional wrapper class name. */
+  className: PropTypes.string,
+
+  /** Pie chart data.  */
+  data: PropTypes.shape({
+
```
Comment (rangoo94) [inline packages/charts/src/PieChart.js]: unnecessary empty line

File: packages/charts/src/PieChart.js [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+ * @param {string} [props.data.dataItems.className]
+ * @param {string} [props.data.dataItems.color]
+ * @param {boolean} [props.data.dataItems.disabled]
+ * @param {string} [props.data.dataItems.label]
+ * @param {number|string} [props.data.dataItems.value]
+ * @param {string} [props.data.title]
+ * @returns {React.Element}
+ */
+function PieChart (props) {
+  const {
```
Comment (rangoo94) [inline packages/charts/src/PieChart.js]: single line?

File: packages/charts/package.json (hunk 18 above)
Comment (rangoo94) [reply packages/charts/package.json]: `0.0.0` actually

File: packages/charts/package.json [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+  "description": "UI component for charts",
+  "main": "dist/index.js",
+  "module": "index.js",
+  "jsnext:main": "index.js",
+  "scripts": {
+    "test": "node ../../scripts/runTests.js --only charts",
+    "prepublishOnly": "npm run build",
+    "build": "node ../../scripts/buildPackages.js --only charts"
+  },
+  "author": "Public in Motion GmbH",
```
Comment (rangoo94) [inline packages/charts/package.json]: copy `publishConfig` from newer packages

File: packages/charts/package.json [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+    "url": "https://github.com/PublicInMotionGmbH/ui-kit/issues"
+  },
+  "license": "MIT",
+  "peerDependencies": {
+    "prop-types": "^15.6.1",
+    "react": "^16.2.0",
+    "react-dom": "^16.2.0"
+  },
+  "dependencies": {
+    "@talixo/checkbox": "^1.0.0",
```
Comment (rangoo94) [inline packages/charts/package.json]: move `dependencies both to peer and dev

**Pull request merged**